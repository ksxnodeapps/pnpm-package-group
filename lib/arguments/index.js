'use strict'
const {usage, example, options, env} = require('./data')

module.exports = (function create (yargs, current, ...next) {
  return current
    ? create(yargs.example(current.cmd, current.desc), ...next)
    : yargs
})(
  require('yargs').usage(usage).options(options).env(env),
  ...example
).help()
