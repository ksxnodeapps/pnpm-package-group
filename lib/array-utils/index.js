'use strict'

const concatReducer = (prev, current) =>
  [...prev, ...current]

module.exports = {
  concatReducer
}
