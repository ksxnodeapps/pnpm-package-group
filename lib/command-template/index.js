'use strict'
const format = require('string-template')

const createContext = (list) => ({
  list,
  argv: list,
  arguments: list,
  pkgs: list,
  packages: list
})

const createString = (tmpl, list) =>
  format(tmpl, createContext(list.join(' ')))

const createArgv = (tmpl, list) =>
  createString(tmpl, list).split(/\s+/)

module.exports = Object.assign(createArgv, {
  createContext,
  createString,
  createArgv
})
