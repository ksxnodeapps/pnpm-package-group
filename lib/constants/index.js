'use strict'
const globalPath = '((GLOBAL))'
module.exports = {globalPath}
