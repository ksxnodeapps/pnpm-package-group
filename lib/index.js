'use strict'
const path = require('path')
const {spawnSync} = require('child_process')
const fs = require('fs')
const process = require('process')
const ramda = require('ramda')
const {_: yamlPaths, ...argvOpts} = require('./arguments').argv
const {exit} = require('./exit-status-names')
const {fetchList} = require('./input')
const constants = require('./constants')
const createArgv = require('./command-template')
const stdio = argvOpts.quiet || argvOpts.quietPnpm ? undefined : 'inherit'
const stepInfo = argvOpts.quiet || argvOpts.quietStep ? () => {} : info

fetchList(yamlPaths)
  .catch(error => exit.withMessage('schema', error.message))
  .then(main)

function main (list) {
  const [
    [{data: globalPackages}],
    localList
  ] = ramda.partition(x => x.path === constants.globalPath, list)

  let globalExitStatus, localExitStatus
  stepInfo('BEGIN')

  if (globalPackages.length) {
    stepInfo('BEGIN GLOBAL')
    globalExitStatus = installGlobalPackages(globalPackages)
    stepInfo('END GLOBAL')
  } else {
    stepInfo('No global packages. Skipping.')
  }

  if (localList.length) {
    stepInfo('BEGIN LOCAL')
    localExitStatus = installLocalPackages(localList)
    stepInfo('END LOCAL')
  } else {
    stepInfo('No local packages. Skipping.')
  }

  const finalExitStatus = (globalExitStatus << 2) | (localExitStatus << 4)
  stepInfo('Status', {global: globalExitStatus, local: localExitStatus, final: finalExitStatus})
  stepInfo('END')
  return process.exit(finalExitStatus)
}

function installGlobalPackages (packages) {
  stepInfo('Installing global packages')
  stepInfo('Packages: ' + packages.join(', '))

  return spawnPnpm(
    createArgv(argvOpts.globalSyntax, packages)
  ).status
}

function installLocalPackages (list) {
  const {writeFileSync} = require('fs-force')
  const pkgjson = require('./package-json-template').json
  const targetDirectory = path.resolve(process.cwd(), argvOpts.local)
  let finalExitStatus = 0
  let total = 0
  let success = 0
  let failure = 0
  stepInfo('Installing local packages')

  for (const item of list) {
    const container = path.resolve(targetDirectory, item.path, argvOpts.packagesLocation)
    const pkgjsonfile = path.resolve(container, 'package.json')
    stepInfo('Location: ' + item.path)
    stepInfo('Packages: ' + item.data.join(', '))
    ++total

    if (!fs.existsSync(pkgjsonfile)) {
      try {
        writeFileSync(pkgjsonfile, pkgjson)
      } catch (error) {
        ++failure
        console.error(error)
        continue
      }
    }

    const cmdArgv = createArgv(argvOpts.localSyntax, item.data)
    const {status} = spawnPnpm(cmdArgv, {cwd: container})

    finalExitStatus += status
    if (status) {
      ++failure
    } else {
      ++success
    }
  }

  stepInfo('Steps', {total, success, failure})
  stepInfo('Status', {finalExitStatus})
  return finalExitStatus
}

function spawnPnpm (cmdArgv, options = {}) {
  const result = spawnSync(argvOpts.pnpm, cmdArgv, {stdio, shell: argvOpts.shell, ...options})
  const {status, signal, error} = result
  stepInfo('Step Result', {status, signal, error})
  return result
}

function info (...args) {
  console.info('[INFO]', ...args)
}
