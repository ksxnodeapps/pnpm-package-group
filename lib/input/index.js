'use strict'

const fetchTree = async yamlPaths => yamlPaths.length
  ? require('../object-utils').merge(...yamlPaths.map(require('../yaml-property')))
  : require('js-yaml').safeLoad(await require('get-stdin')())

const fetchList = async yamlPaths =>
  require('../tree-to-list')(await fetchTree(yamlPaths))

module.exports = {
  fetchTree,
  fetchList
}
