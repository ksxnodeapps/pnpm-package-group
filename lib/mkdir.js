module.exports = require('fs-force').mkdirSync
