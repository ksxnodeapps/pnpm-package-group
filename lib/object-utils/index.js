'use strict'
const deepmerge = require('deepmerge')

const entriesReducer = (obj, [key, val]) =>
  Object.assign(obj, {[key]: val})

const merge = (...args) =>
  args.reduce((a, b) => deepmerge(a, b), {})

const nestedObject = (value, path) => {
  if (path.length === 0) return value
  const [first, ...rest] = path
  return {[first]: nestedObject(value, rest)}
}

const filterEntries = (object, filter) => Object
  .entries(object)
  .filter(pair => filter(pair, object))
  .reduce(entriesReducer, {})

module.exports = {
  entriesReducer,
  merge,
  nestedObject,
  filterEntries
}
