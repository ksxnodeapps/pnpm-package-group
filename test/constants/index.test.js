'use strict'

describe('subject', () => {
  it('stays unchanged', () => {
    expect(
      require('../../lib/constants')
    ).toMatchSnapshot()
  })
})
