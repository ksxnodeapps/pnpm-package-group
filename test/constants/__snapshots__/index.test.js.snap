// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subject stays unchanged 1`] = `
Object {
  "globalPath": "((GLOBAL))",
}
`;
